File: Fortran/3_sendReceive/solution/d/send_receive.f90
program sendReceive
    use mpi
    implicit none

    integer, parameter :: maxElements = 1048576

    integer :: myRank, numProcs, ierr
    integer, allocatable, dimension(:) :: data, tmpdata
    integer :: nextRank, prevRank, numElements
    integer :: requests(2)

    allocate(data(maxElements))
    allocate(tmpdata(maxElements))
    if (.not.allocated(data) .or. .not.allocated(tmpdata)) then
        print *,"Not enough memory"
        stop 1
    end if
    ! Note: Data remains uninitialized, as this is just an example

    ! MPI Initialization
    call MPI_INIT(ierr)
    call MPI_COMM_RANK(MPI_COMM_WORLD, myRank, ierr)
    call MPI_COMM_SIZE(MPI_COMM_WORLD, numProcs, ierr)

    if (numProcs /= 2) then
        print *,"Error: This program can only run with 2 MPI tasks!"
        call MPI_ABORT(MPI_COMM_WORLD, 1, ierr)
        stop 1
    end if

    nextRank = mod(myRank+1, numProcs)
    prevRank = mod(myRank-1+numProcs, numProcs)

    do numElements = 4096, maxElements, 4096
        print '("Process ",I0," sends ",I0," elements of data now")', &
            myRank, numElements
        call MPI_ISEND(data, numElements, MPI_INTEGER, nextRank, 0, &
            MPI_COMM_WORLD, requests(1), ierr)
        print '("Process ",I0," receives ",I0," elements of data now")', &
            myRank, numElements
        ! Cannot use data to receive messages since it might still be in use
        ! by the sending operation. That's why we use a temporary buffer.
        call MPI_IRECV(tmpdata, numElements, MPI_INTEGER, MPI_ANY_SOURCE, 0, &
            MPI_COMM_WORLD, requests(2), ierr)
        call MPI_WAITALL(2, requests, MPI_STATUSES_IGNORE, ierr)
        ! Copy the temporary receive buffer to data
        data(1:numElements) = tmpdata(1:numElements)
        print '("Process ",I0," is done with ",I0," elements of data")', &
            myRank, numElements
    end do

    deallocate(data)

    call MPI_FINALIZE(ierr)
end program sendReceive

File: C/3_sendReceive/solution/d/send_receive.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <mpi.h>

int main(int argc, char **argv)
{
    int maxElements = 1048576;

    int *data = (int *)malloc(sizeof(int) * maxElements);
    int *tmpdata = (int *)malloc(sizeof(int) * maxElements);
    if (data == NULL || tmpdata == NULL)
    {
        printf("Not enough memory\n");
        return -1;
    }
    else
    {
        // Initialize with random data for this example
        srand(42);
        for (int i = 0; i < maxElements; ++i)
        {
            data[i] = rand();
        }
    }

    int worldRank, worldSize;
    // MPI Initialization
    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &worldSize);
    MPI_Comm_rank(MPI_COMM_WORLD, &worldRank);

    if (worldSize != 2)
    {
        printf("This program can only be started with 2 MPI processes\n");
        MPI_Abort(MPI_COMM_WORLD, 1);
        return 1;
    }

    int nextRank = (worldRank + 1) % worldSize;
    int prevRank = (worldRank - 1 + worldSize) % worldSize;
    int numElements;

    for (numElements = 4096; numElements < maxElements; numElements += 4096)
    {
        MPI_Status statuses[2];
        MPI_Request requests[2];

        printf("Rank %i sends %i elements of data now\n",
            worldRank, numElements);
        MPI_Isend(data, numElements, MPI_INT, nextRank, 0, MPI_COMM_WORLD,
            &requests[0]);
        printf("Rank %i receives %i elements of data now\n",
            worldRank, numElements);
        // Cannot use data to receive messages since it might still be in use
        // by the sending operation. That's why we use a temporary buffer.
        MPI_Irecv(tmpdata, numElements, MPI_INT, prevRank, 0,
            MPI_COMM_WORLD, &requests[1]);
        MPI_Waitall(2, requests, statuses);
        // Sent to previous rank, receive from next
        printf("Rank %i sends %i elements of data in backward direction\n",
            worldRank, numElements);
        MPI_Isend(tmpdata, numElements, MPI_INT, prevRank, 0, MPI_COMM_WORLD,
            &requests[0]);
        printf("Rank %i receives %i elements of data now\n",
            worldRank, numElements);
        MPI_Irecv(data, numElements, MPI_INT, nextRank, 0, MPI_COMM_WORLD,
            &requests[1]);
        MPI_Waitall(2, requests, statuses);

        printf("Rank %i is done with %i elements of data\n",
            worldRank, numElements);
        // Copy the temporary receive buffer to data
        memcpy(data, tmpdata, numElements*sizeof(int));
    }

    MPI_Finalize();
    return 0;
}
